src/erlcloud_aws.erl

```diff
@@ -989,22 +996,41 @@ sign_v4_headers(Config, Headers, Payload, Region, Service) ->
 -spec sign_v4(atom(), list(), aws_config(), headers(), string() | binary(), string(), string(), list()) -> headers().
 sign_v4(Method, Uri, Config, Headers, Payload, Region, Service, QueryParams) ->
     Date = iso_8601_basic_time(),
-    {PayloadHash, Headers1} =
-        sign_v4_content_sha256_header( [{"x-amz-date", Date} | Headers], Payload ),
-    Headers2 = case Config#aws_config.security_token of
-                   undefined -> Headers1;
-                   Token -> [{"x-amz-security-token", Token} | Headers1]
+    sign_v4(Method, Uri, Config, Headers, Payload, Region, Service, QueryParams, Date).
+
+-spec sign_v4(atom(), list(), aws_config(), headers(), string() | binary(), string(), string(), list(), string()) -> headers().
+sign_v4(Method, Uri, Config, Headers0, Payload, Region, Service, QueryParams, Date0) ->
+    % use passed-in x-amz-date header or create one
```

Comment:
I find the semantics of this a _little_ strange in that we are passing an explicit date but the x-amz-date command overrides this. I assume this is because of the double-duty this is doing for verifying our signatures (which will likely have the date header coming in via Headers) and generating new signatures (which I am assuming won't have the date header coming in via Headers)

Reply:
I don't know if this helps, but the way it works is: you pass in a date, but if there is a date header also being passed in, that date is used.  If not, the passed-in date is used.

Reply:
I think your 'double-duty' comment is probably correct.